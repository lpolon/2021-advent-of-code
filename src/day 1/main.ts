const parseInput = (input: string): number[] => input.split(/\n/).map(Number);

const countDepthIncreasesFromNumbers = (input: number[]): number => {
  return input.reduce<number>((acc, value, index, arr) => {
    const prevValue = arr[index - 1];
    if (value - prevValue > 0) acc += 1;
    return acc;
  }, 0);
};

export const countDepthIncreases = (input: string) => {
  const values = parseInput(input);
  return countDepthIncreasesFromNumbers(values);
};

export const countDepthIncreasesInThreeMeasurementSlidingWindow = (input: string) => {
  const values = parseInput(input);

  const threeValueSum = values
    .map<number | undefined>((value, index, arr) => {
      const valueMinusOne = arr[index - 1];
      const valueMinusTwo = arr[index - 2];

      if (!valueMinusOne || !valueMinusTwo || !value) {
        return;
      }
      return valueMinusTwo + valueMinusOne + value;
    })
    .filter((e: number | undefined): e is number => Boolean(e));

  return countDepthIncreasesFromNumbers(threeValueSum);
};
