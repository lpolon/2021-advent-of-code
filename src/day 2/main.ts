const subDirectionTup = ['forward', 'down', 'up'] as const;
type SubDirection = typeof subDirectionTup[number];
type SubInstruction = {
  direction: SubDirection;
  distance: number;
};

const isValidDirection = (d: string): d is SubDirection => {
  return subDirectionTup.includes(d);
};

export const parseInput = (input: string): SubInstruction[] =>
  input
    .split(/\n/)
    .map((i) => i.split(/\s/))
    .map(([direction, distance]) => {
      if (!isValidDirection(direction)) {
        return;
      }
      return {
        direction: direction,
        distance: Number(distance),
      };
    })
    .filter((s: SubInstruction | undefined): s is SubInstruction => Boolean(s));
