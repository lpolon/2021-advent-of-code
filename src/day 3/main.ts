type BitValues = '0' | '1';

type BitFrequencyHt = {
  [Key in BitValues]: number;
};

const parseInput = (input: string): string[] => input.split(/\n/);

const isValidBit = (bit: string): bit is '0' | '1' => bit === '0' || bit === '1';

export const getModalBitByIndex = (input: string, index: number): '1' | '0' => {
  const bitFrequencyHt = parseInput(input).reduce<BitFrequencyHt>(
    (acc, curr) => {
      const currentBit = curr[index];
      if (!isValidBit(currentBit)) return acc;
      acc[currentBit] += 1;
      return acc;
    },
    { '0': 0, '1': 1 },
  );

  if (bitFrequencyHt[0] > bitFrequencyHt[1]) {
    return '0';
  }
  return '1';
};

export const getBitNumberLength = (input: string): number => {
  return parseInput(input)[0].length;
};

export const calcGammaRate = (input: string): number => {
  const bitNumberLength = getBitNumberLength(input);
  const gammaBitNumber = Array(bitNumberLength)
    .fill(null)
    .map((_, index) => {
      return getModalBitByIndex(input, index);
    })
    .join('');

  return parseInt(gammaBitNumber, 2);
};

export const calcEpsilonRate = (input: string): number => {
  const bitNumberLength = getBitNumberLength(input);
  const epsilonBitNumber = Array(bitNumberLength)
    .fill(null)
    .map((_, index) => {
      return getModalBitByIndex(input, index) === '1' ? '0' : '1';
    })
    .join('');

  return parseInt(epsilonBitNumber, 2);
};

export const calcPowerConsumption = (input: string): number => {
  const epsilonRate = calcEpsilonRate(input);
  const gammaRate = calcGammaRate(input);

  return epsilonRate * gammaRate;
};

// Day 3 - 2 -- i should split these files

/*
A grande diferença entre a parte 1 e a parte 2 é que o array muda a cada iteração na parte dois. utiliza-se o array filtrado.
*/
