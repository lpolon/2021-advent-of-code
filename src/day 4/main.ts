/*
approach heavily inspired by
https://github.com/guyroyse/advent-of-code-2021/blob/main/lib/dec.04.js

I liked the class approach because it handles the mutation of the boards in a clean fashion.
The fact that the board inside the Board class is flattened is pretty neat. I definitely would not have thought of it myself. I felt like a cheated in this one, but it was enlightening.
*/

import { transpose } from './helpers';

type DrawnNumbers = number[];

export class Cell {
  private _value: number;
  private _isMarked: boolean;

  constructor(number: number) {
    this._value = number;
    this._isMarked = false;
  }

  get value() {
    return this._value;
  }

  get isMarked() {
    return this._isMarked;
  }

  mark() {
    this._isMarked = true;
  }
}

export class Board {
  // TODO: board n√£o flatten
  // columns
  // rows
  private _flattenedCells: Cell[];
  private _rows: Cell[][];
  private _columns: Cell[][];

  constructor(values: string) {
    const cells = values.split(/\n/).map<Cell[]>((row) =>
      row
        .split(/\s/)
        .filter(Boolean)
        .map(Number)
        .map((cellValue) => new Cell(cellValue)),
    );

    this._flattenedCells = cells.flat();
    this._rows = cells;
    this._columns = transpose(cells);
  }

  public mark(number: number): void {
    this._flattenedCells.find((cell) => cell.value === number)?.mark();
  }
  public get isWinner(): boolean {
    return (
      this._rows.some((row) => row.every((cell) => cell.isMarked)) ||
      this._columns.some((column) => column.every((cell) => cell.isMarked))
    );
  }
}

const parseBoards = (boards: string[]): Board[] => {
  return boards.map((board) => new Board(board));
};

export const parseInput = (input: string): { drawnNumbers: DrawnNumbers; bingoBoards: Board[] } => {
  const [drawnNumbersStr, ...boards] = input.split(/\n\n/);
  const drawnNumbers = drawnNumbersStr.split(/,/).map(Number);
  const bingoBoards = parseBoards(boards);
  return {
    drawnNumbers,
    bingoBoards,
  };
};

const updateBoards = (number: number, boards: Board[]): void => {};
