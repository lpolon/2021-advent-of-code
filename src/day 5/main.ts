type Coords = {
  x1: number;
  x2: number;
  y1: number;
  y2: number;
};

export const parseInput = (input: string): Coords[] => {
  const coords: Coords[] = [];

  for (const coordSet of input.split(/\n/)) {
    const matchResult = coordSet.match(/(^.),(.)\s->\s(.),(.)/);
    if (!matchResult) {
      continue;
    }
    const [,x1, y1, x2, y2] = matchResult.map(Number);
    coords.push({ x1, y1, x2, y2 });
  }
  return coords;
};
