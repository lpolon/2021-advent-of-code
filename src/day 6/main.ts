export const parseInput = (input: string): number[] => {
  return input.split(',').map(Number);
};

export const updateLanternfishState = (fishes: number[]): number[] => {
  const newFishes: number[] = [];
  const updatedFishes = fishes.map((fish) => {
    if (fish === 0) {
      newFishes.push(8);
      return 6;
    }
    return fish - 1;
  });
  return [...updatedFishes, ...newFishes];
};

export const countFishes = (days: number, fishes: number[]): number => {
  if (days === 0) return fishes.length;
  const updatedFishes = updateLanternfishState(fishes);
  return countFishes(days - 1, updatedFishes);
};
