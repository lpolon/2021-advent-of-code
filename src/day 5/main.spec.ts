import fs from 'fs';
import path from 'path';
import { parseInput } from './main';

const fetchExampleInput = () =>
  fs.readFileSync(path.resolve(__dirname, './example.input.txt'), 'utf-8');

/*
[
      { x1: 0, y1: 9, x2: 5, y2: 9 },
      { x1: 8, y1: 0, x2: 0, y2: 8 },
      { x1: 9, y1: 4, x2: 3, y2: 4 },
      { x1: 2, y1: 2, x2: 2, y2: 1 },
      { x1: 7, y1: 0, x2: 7, y2: 4 },
      { x1: 6, y1: 4, x2: 2, y2: 0 },
      { x1: 0, y1: 9, x2: 2, y2: 9 },
      { x1: 3, y1: 4, x2: 1, y2: 4 },
      { x1: 0, y1: 0, x2: 8, y2: 8 },
      { x1: 5, y1: 5, x2: 8, y2: 2 }
    ]
*/

test('parse input works', () => {
  const input = fetchExampleInput();
  const output = parseInput(input);
  expect(output).toMatchSnapshot();
});
