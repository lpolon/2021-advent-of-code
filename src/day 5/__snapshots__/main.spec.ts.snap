// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`parse input works 1`] = `
Array [
  Object {
    "x1": 0,
    "x2": 5,
    "y1": 9,
    "y2": 9,
  },
  Object {
    "x1": 8,
    "x2": 0,
    "y1": 0,
    "y2": 8,
  },
  Object {
    "x1": 9,
    "x2": 3,
    "y1": 4,
    "y2": 4,
  },
  Object {
    "x1": 2,
    "x2": 2,
    "y1": 2,
    "y2": 1,
  },
  Object {
    "x1": 7,
    "x2": 7,
    "y1": 0,
    "y2": 4,
  },
  Object {
    "x1": 6,
    "x2": 2,
    "y1": 4,
    "y2": 0,
  },
  Object {
    "x1": 0,
    "x2": 2,
    "y1": 9,
    "y2": 9,
  },
  Object {
    "x1": 3,
    "x2": 1,
    "y1": 4,
    "y2": 4,
  },
  Object {
    "x1": 0,
    "x2": 8,
    "y1": 0,
    "y2": 8,
  },
  Object {
    "x1": 5,
    "x2": 8,
    "y1": 5,
    "y2": 2,
  },
]
`;
